capstone/src/main/java/com/google/sps/Objects/User.java

```diff
@@ -5,28 +5,31 @@
 public final class User {
 
   private final String userId;
-  private final String name;
-  private final String firstName;
-  private final String lastName;
+  private String name;
+  private String firstName;
+  private String lastName;
   private final String email;
-  private final String phoneNumber;
+  private String phoneNumber;
   private final ArrayList<Badge> badges;
   private final ArrayList<String> groups;
   private final ArrayList<String> interests;
 
   public User(String userId, String firstName, String lastName, String email) {
     this.userId = userId;
     this.name = firstName + " " + lastName;
-    this.firstName = firstname;
+    this.firstName = firstName;
     this.lastName = lastName;
     this.email = email;
+    this.badges = new ArrayList<>();
+    this.groups = new ArrayList<>();
```

Comment:
This will be a similar comment to the one I made in Post.java about addNewLike, except this is about addGroup. Github is not allowing me to comment on addGroup however since you didn't make changes to it in this PR. I understand you may not have been the person to make this file, however I think you can apply my comment above to this method as well.

Specifically, based on the comment "Adds a unique groupId string to the user's list of groups."
What is ensuring that a unique groupId is being added? This method should take care of that rather than it depend on the ID being passed in to be unique.

Reply:
Hmm that's a good point. I'm a little confused because in my mind, the User class itself (and I think all of these classes) would only be used after retrieving things from the Datastore. Because we retrieve data from the database, use that data to create a User, Group, Post, etc object, and then convert that object to JSON which is the response to the client. So wouldn't ensuring the values are unique be the job of the server side code?

Reply:
The database would make sure there aren't two Users with identical IDs, such that the newest user with the same ID would override the existing one. However, with respect to a User's list of Groups, it wouldn't know whether or not that list needs to be unique - it just stores the list you give it. Sure, the database would have unique Groups, but at some point you need to do a check that you aren't doing something with a Group that already exists, whether that be when you add it (ideal), when you get the list of Groups, when you decide you want to do something with that list of Groups. Imagine we don't ensure the list of Groups is unique, so we store the list where we have a Group that shows up twice. Then on a User's profile we want to display all of the Groups they are in. If you haven't ensure the data is unique, then here the Group would show up twice unless you do a check such that you make sure any Group that is being displayed isn't already displayed.

Ideally the purpose of the Servlets should be just retrieving/storing data. These objects act as the middle layer between the database and frontend - they should be used to structure the data when we retrieve the data and modify the data until we are ready to store it. Otherwise, we are just using these objects to send in a RESPONSE and not utilizing the objects to their best potential. 

I understand we haven't exactly been doing it like so for the portfolio and it might seem trivial, but the flow for adding a new group should be something like so:

[User] User clicks on Join Group button ->
[JavaScript] Join Group button performs a POST request ->
[Servlet] User entity is retrieved from database using user ID from the request ->
[Servlet] User object created from the User entity ->
[User Object] Add group to User object's set of Groups. If Group already exists, it is not added ->
[Servlet] Update User in database with the new set of Groups